posts/2024-09-05-impl-trait-capture-rules.md

```diff
@@ -0,0 +1,211 @@
+---
+layout: post
+title: "Changes to impl Trait in Rust 2024"
+author: Niko Matsakis
+team: the language team <https://www.rust-lang.org/governance/teams/lang>
+---
+This blog post describes some small but significant changes with (return position) `impl Trait` capture rules that are coming in Rust 2024. The goal of these changes is to simplify how `impl Trait` works to better match what people want while also giving a flexible syntax that allows users to have full control when needed.
+
+## TL;DR
+
+* Return-position impl Trait in Rust 2024 can now reference all fn arguments by default.
+* We are introducing a new syntax (`+ use<>`) that you can use to specify precisely which generic parameters can appear in the hidden type.
+
+## Background: return position impl trait
+
+This blog post concerns *return position `impl Trait`*, such as the following example:
+
+```rust
+fn process_datums(
+    datums: &[Datum]
+) -> impl Iterator<Item = ProcessedDatum> {
+    datums
+        .iter()
+        .map(|datum| datum.process())
+}
+```
+
+The use of `-> impl Iterator` in return position here means that the functions "some kind of iterator". The actual type will be determined by the compiler based on the function body. It is called the "hidden type" because callers do not get to know exactly what it is, they have to code against the `Iterator` trait. However, at code generation time, the compiler will generate code based on the actual precise type, which ensures that callers are fully optimized.
```

Comment:
```suggestion
The use of `-> impl Iterator` in return position here means that the functions "some kind of iterator". The actual type will be determined by the compiler based on the function body. It is called the "hidden type" because callers do not get to know exactly what it is; they have to code against the `Iterator` trait. However, at code generation time, the compiler will generate code based on the actual precise type, which ensures that callers are fully optimized.
```